### PR TITLE
Fix inhibit race

### LINKED ISSUE #1014: Race in inhibitor shutdown
Same testing round as #1013, another race that was detected.

```
11:43:04 a1 | ==================
11:43:04 a1 | WARNING: DATA RACE
11:43:04 a1 | Write at 0x00c4202984d0 by goroutine 90:
11:43:04 a1 |   github.com/prometheus/alertmanager/inhibit.(*Inhibitor).Stop()
11:43:04 a1 |       /home/brancz/go/src/github.com/prometheus/alertmanager/inhibit/inhibit.go:111 +0xe9
11:43:04 a1 |   main.main.func8()
11:43:04 a1 |       /home/brancz/go/src/github.com/prometheus/alertmanager/cmd/alertmanager/main.go:270 +0x440
11:43:04 a1 |   main.main.func9()
11:43:04 a1 |       /home/brancz/go/src/github.com/prometheus/alertmanager/cmd/alertmanager/main.go:332 +0x68
11:43:04 a1 | Previous read at 0x00c4202984d0 by goroutine 40:
11:43:04 a1 |   github.com/prometheus/alertmanager/inhibit.(*Inhibitor).Run()
11:43:04 a1 |       /home/brancz/go/src/github.com/prometheus/alertmanager/inhibit/inhibit.go:79 +0x173
11:43:04 a1 | Goroutine 90 (running) created at:
11:43:04 a1 |   main.main()
11:43:04 a1 |       /home/brancz/go/src/github.com/prometheus/alertmanager/cmd/alertmanager/main.go:325 +0x22ea
11:43:04 a1 | Goroutine 40 (running) created at:
11:43:04 a1 |   main.main.func8()
11:43:04 a1 |       /home/brancz/go/src/github.com/prometheus/alertmanager/cmd/alertmanager/main.go:286 +0x7a3
11:43:04 a1 |   main.main()
11:43:04 a1 |       /home/brancz/go/src/github.com/prometheus/alertmanager/cmd/alertmanager/main.go:291 +0x1aa3
11:43:04 a1 | ==================
```

This race only showed when shutting down the Alertmanager, and it seems there is a race between [closing and clearing the `stopc` channel](https://github.com/prometheus/alertmanager/blob/9f5f4b2a516d35cfaf196530b277f1d109254569/inhibit/inhibit.go#L110-L111) and possibly [reading from it](https://github.com/prometheus/alertmanager/blob/master/inhibit/inhibit.go#L79). If I understand correctly, this can cause the Alertmanager to deadlock as receiving from a nil channel blocks forever, on shutdown of the inhibitor which happens at shutdown and reload of the Alertmanager.

@stuartnelson3 @fabxc

### PR DESCRIPTION
**Pull request by brancz**
Fixes #1014

@fabxc @stuartnelson3

### CODE REVIEW

File: inhibit/inhibit.go [hunk 1]
```diff
@@ -50,33 +52,26 @@ func NewInhibitor(ap provider.Alerts, rs []*config.InhibitRule, mk types.Marker)
...
+func (ih *Inhibitor) runGC(ctx context.Context) {
 	for {
 		select {
 		case <-time.After(15 * time.Minute):
 			for _, r := range ih.rules {
 				r.gc()
 			}
-		case <-ih.stopc:
+		case <-ctx.Done():
 			return
```
Comment (josedonizetti) [inline inhibit/inhibit.go]: Think you should return `ctx.Err()`

File: inhibit/inhibit.go [hunk 2]
```diff
@@ -50,33 +52,26 @@ func NewInhibitor(ap provider.Alerts, rs []*config.InhibitRule, mk types.Marker)
...
-
+func (ih *Inhibitor) run(ctx context.Context) {
 	it := ih.alerts.Subscribe()
 	defer it.Close()
 
 	for {
 		select {
-		case <-ih.stopc:
+		case <-ctx.Done():
 			return
```
Comment (josedonizetti) [inline inhibit/inhibit.go]: Also `return ctx.Err()` here

File: inhibit/inhibit.go [hunk 3]
```diff
@@ -98,17 +93,41 @@ func (ih *Inhibitor) Run() {
...
+		ctx context.Context
+	)
+
+	ctx, ih.cancel = context.WithCancel(context.Background())
+	gcCtx, gcCancel := context.WithCancel(ctx)
+	runCtx, runCancel := context.WithCancel(ctx)
+
+	g.Add(func() error {
+		ih.runGC(gcCtx)
+		return nil
```
Comment (josedonizetti) [inline inhibit/inhibit.go]: If you return the context error, you can return it here. And the same for the ih.run below.

File: inhibit/inhibit.go (hunk 1 above)
Comment (brancz) [reply inhibit/inhibit.go]: `ctx.Err()` just returns why a context was canceled or deadline exceeded. It doesn't actually tell us anything here.

```
        // If Done is not yet closed, Err returns nil.
        // If Done is closed, Err returns a non-nil error explaining why:
        // Canceled if the context was canceled
        // or DeadlineExceeded if the context's deadline passed.
        // After Err returns a non-nil error, successive calls to Err return the same error.
        Err() error
```

Comment (josedonizetti) [reply inhibit/inhibit.go]: true, but if non nil isn't worth logging?

Comment (brancz) [reply inhibit/inhibit.go]: I don't think so, because we explicitly use the context to cancel the execution in the `Inhibitor.Stop` method.

Comment (josedonizetti) [reply inhibit/inhibit.go]: Gotcha 👍

Comment (fabxc) [reply inhibit/inhibit.go]: Yes, the error is always `!=` nil if the context was canceled. If we expect this to happen, there's generally no point in checking.

**Pull request merged**